Add v2 unit test for loss functions and update loss/metrics

am2145 opened the pull request:
## Description
This PR adds a unit test for the loss functions defined in `loss.py`. The tests are not exhaustive, but include a test for each loss function available and check for basic functionality such as expected dimensions, reduction, and simple parameters. Some loss function definitions in `loss.py` are modified to address some issues. These are: 
1) BoundedMSELoss is modified to have the lt and gt masks applied to the targets in addition to preds. 
2) MulticlassMCCLoss is modified to have summations such that the end vector is a scalar. This should match the implementation in v1. 
3) BinaryDirichletLoss and MulticlassDirichletLoss have had their forward call modified to pass to DirichletMixin as the loss was previously returning a NoneType.  

`metrics.py` has also been updated to have metrics inherit from loss functions where possible. Specifically SID and Wasserstein metrics now inherit from the corresponding loss functions. The MCCmetric has been replaced with BinaryMCC and MulticlassMCC metrics to reflect the updated loss functions in loss.py as well. BoundedMixin has been updated to apply the lt and gt masks applied to targets as well, matching the change for BoundedMSELoss in `loss.py`

## Questions
1) The changes made to MulticlassMCCLoss and the DirichletLoss functions work, but are they in line with what was intended for the v2 framework?

davidegraff reviewed (commented):
beyond the open questions about calls to `super()`, my only comment is to using `torch.testing` in test files

am2145 commented on chemprop/v2/models/loss.py:
```diff
@@ -257,16 +257,15 @@ def forward(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
         y_one_hot = torch.eye(preds.shape[2], device=preds.device)[targets.long()]
 
-        return super().forward(preds, y_one_hot, mask)
+        return super(LossFunction, self).forward(preds, y_one_hot, mask)
```
No reason as it just happened to be what worked first. I have changed this now to use `DirichletMixin.forward` and the result is the same, so we should be good to go here.

am2145 commented:
> beyond the open questions about calls to `super()`, my only comment is to using `torch.testing` in test files

Thank you for reviewing the PR and the feedback. I have refactored the tests to use `torch.testing` as requested and also updated the DirichletLoss functions. I have also rebased on Hao-Wei and Xiaorui's previous PRs that have been merged in v2/dev already. Please let me know if there are any further changes that are required.

davidegraff commented on chemprop/v2/models/loss.py:
```diff
@@ -257,16 +257,16 @@ def forward(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
-
-        return super().forward(preds, y_one_hot)
+        
+        return DirichletMixin.forward(self, preds, y_one_hot, *args)
```
```suggestion
        return super().forward(preds, y_one_hot, *args)
```
If you're using the `super`'s method, then it's standard to call it using the `super()` object

am2145 replied:
When I try this, the forward method from `LossFunction` gets called first. This means that line 52 returns an error as the Loss is `NoneType` because it has not gone through DirichletMixin yet. What's the correct way to use `super()` here?

davidegraff commented on chemprop/v2/models/loss.py:
```diff
@@ -257,16 +257,16 @@ def forward(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
         y_one_hot = torch.eye(preds.shape[2], device=preds.device)[targets.long()]
 
-        return super().forward(preds, y_one_hot, mask)
+        return DirichletMixin.forward(self, preds, y_one_hot, mask)
```
see above

davidegraff commented on chemprop/v2/tests/unit/test_loss_functions.py:
```diff
@@ -0,0 +1,484 @@
+    """
+    binary_dirichlet_loss = BinaryDirichletLoss(v_kl = v_kl)
+    loss = binary_dirichlet_loss(preds, targets, mask, w_s, w_t, lt_mask, gt_mask)
+    print(loss)
```
```suggestion

```

davidegraff commented on chemprop/v2/tests/unit/test_loss_functions.py:
```diff
@@ -0,0 +1,484 @@
+    Test on the dirichlet loss function for classification
+    for dimension errors.
+    """
+    with pytest.raises(RuntimeError):
+        binary_dirichlet_loss = BinaryDirichletLoss()
+        binary_dirichlet_loss(preds, targets, mask, w_s, w_t, lt_mask, gt_mask)
```
good test- should we think about catching this and throwing a more descriptive error?

am2145 replied:
I adapted this test from v1. Do we want to print a wrong input dimension error?

davidegraff replied:
we decided to ignore my comment

am2145 commented:
I have swapped the inheritance of the DirichletLoss functions to have `DirichletMixin` first and then have their forward method use `super()` as we discussed in the meeting. I also deleted the print statement and rebased off of v2/dev so everything should now be up to date. Thanks again for all the help.

davidegraff reviewed (approved):
thanks for the great work @am2145 !
